### PR TITLE
Add Gumbel rounding 

### PR DESCRIPTION
**Pull request by haihabi**
Refactor quantizer in GPTQ
Adding Gumbel Rounding quantizers

### CODE REVIEW

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/uniform_gumbel.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                                      auxvar_tensor: tf.Variable,
+                                      range_min: tf.Tensor,
+                                      range_max: tf.Tensor,
+                                      n_bits: int) -> tf.Tensor:
+    """
+    Quantize a tensor according to given range (min, max) and number of bits.
+
+    Args:
+        tensor_data: Tensor values to quantize.
+        auxvar_tensor:
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/uniform_gumbel.py:38]: auxvar_tensor comment

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/uniform_gumbel.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                                                         self.num_bits)
+            return q_tensor
+        else:
+            raise NotImplemented
+            return gumbel_rounding_uniform_quantizer(inputs, auxvar_hat,
+                                                     ptq_max_range,
+                                                     ptq_min_range,
+                                                     self.num_bits)
+
+    def calc_quant_config(self, layer):
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/uniform_gumbel.py:199]: get_quant_config seems better
return type hint

File: model_compression_toolkit/gptq/common/gptq_config.py [hunk 3]
```diff
@@ -44,12 +55,26 @@ def __init__(self,
...
         self.log_function = log_function
         self.train_bias = train_bias
+        self.quantization_parameters_learning = quantization_parameters_learning
+        self.temperature_learning = temperature_learning
+        self.rounding_type = rounding_type
+        self.sam_optimization = sam_optimization
+        self.rho = rho
         self.lsb_change_per_bit_width = lsb_change_per_bit_width
+
+    def is_gumbel(self) -> bool:
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/common/gptq_config.py:79]: Missing comment

File: model_compression_toolkit/gptq/keras/quantizer/kernel_functions.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    for w in weights_list:
+        if KERNEL in w.name:
+            return w
+    raise Exception("Can't find kernel variable")
+
+
+def threshold_reshape(threshold_tensor: tf.Tensor, input_w: tf.Tensor, in_quantization_axis: int) -> tf.Tensor:
+    """
+    This function take an threshold tensor and re-aline it to the weight tensor channel axis.
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/kernel_functions.py:36]: takes a

File: model_compression_toolkit/gptq/keras/gptq_training.py [hunk 5]
```diff
@@ -92,47 +104,104 @@ def build_gptq_model(self):
...
+        # ----------------------------------------------
+        if self.gptq_config.train_bias:
+            if self.gptq_config.optimizer_rest is None:
+                common.Logger.error("To enable bias micro training ")
+            common.Logger.info("Starting Bias Training")
+            self.micro_training_loop(representative_data_gen, int(0.2 * self.gptq_config.n_iter),
+                                     self.flattened_bias_weights)
+
+    def micro_training_loop(self, data_function, n_iteration, variable2update):
+        """
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/gptq/keras/gptq_training.py:183]: In general, it seems that this function is redundant. The logic of computing and applying the gradients are in train. Maybe we should use a function for both of them as the difference is the number of training iterations.

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/gumbel_softmax.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    This function apply ste on the output of the gumbel softmax.
+    Args:
+        in_prob:A tensor of probability
+
+    Returns: A Tensor of ohe hot vector with STE.
+
+    """
+
+    delta = tf.stop_gradient(select_gumbel(in_prob) - in_prob)
```
[the remark this answers is not in the document]
Comment (haihabi) [reply model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/gumbel_softmax.py:57]: Explain?

File: model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
+# Copyright 2021 Sony Semiconductors Israel, Inc. All rights reserved.
+#
```
[the remark this answers is not in the document]
Comment (haihabi) [reply model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py:2]: Why this is an old file

File: model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+            that changed during GPTQ retraining.
+            Keys must match NodeQuantizationConfig attributes
+
+        """
+        weights = {}
+        for weight, quantizer, quantizer_vars in layer._weight_vars:
+            weights.update({KERNEL: quantizer(weight, training=False, weights=quantizer_vars)})
+
+        quant_config = {'weights_quantization_params': self.weight_quantizer.calc_quant_config(layer)}
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py:213]: Maybe we should use it in the gptq constants

File: model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        return weights, quant_config, {}
+
+    def get_trainable_quantizer_parameters(self):
+        """
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/keras/quantizer/configs/weight_quantizer_gptq_config.py:217]: Type hints (to all methods in this file)

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/gptq/keras/gptq_training.py [hunk 10]
```diff
@@ -92,47 +104,104 @@ def build_gptq_model(self):
...
+        Returns:
+            Loss and gradients.
+        """
+
+        with tf.GradientTape(persistent=True) as tape:
+            y_fxp = self.fxp_model(input_data, training=training)  # running fxp model
+            loss_value = self.gptq_config.loss(y_fxp, in_y_float, self.fxp_weights_list, self.flp_weights_list,
+                                               self.compare_points_mean, self.compare_points_std)
+            if self.gptq_config.is_gumbel() and self.gptq_config.temperature_learning:
+                gumbel_prob = get_gumbel_probability(self.fxp_model)
```
[the remark this answers is not in the document]
Comment (haihabi) [reply model_compression_toolkit/gptq/keras/gptq_training.py:127]: No

File: model_compression_toolkit/gptq/keras/gptq_training.py [hunk 11]
```diff
@@ -92,47 +104,104 @@ def build_gptq_model(self):
...
-            # Run one step of gradient descent by updating
-            # the value of the variables to minimize the loss.
-            self.gptq_config.optimizer.apply_gradients(zip(grads, self.flattened_trainable_weights))
-
-            return loss_value, grads
-
+        if self.gptq_config.temperature_learning:
+            w2train = [*self.flattened_trainable_weights, *self.temperature_weights]
+        else:
+            w2train = self.flattened_trainable_weights
```
[the remark this answers is not in the document]
Comment (haihabi) [reply model_compression_toolkit/gptq/keras/gptq_training.py:151]: Nice catch

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+from model_compression_toolkit.core.keras.quantizer.base_quantizer import BaseTrainableQuantizer
+from model_compression_toolkit.core.common.defaultdict import DefaultDict
+from model_compression_toolkit.core import common
+from model_compression_toolkit.gptq.keras.quantizer import kernel_functions
+from model_compression_toolkit.gptq.keras.quantizer.gumbel_rounding.gumbel_softmax import sample_gumbel
+from model_compression_toolkit.gptq.common import gptq_constants
+
+
+def _init_aux_var(w_shape, m):
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py:28]: Comments and type hints.

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    m_hat = m // 2
+    p = 0.01
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py:30]: Please use a constant.

File: model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    shift = -np.log(-np.log(1 - p))
+    n = np.random.randn(*[m, *w_shape]) * np.sqrt(np.power(np.pi, 2) / 6)
+    n[m_hat, :] += shift
+    return n
+
+
+def _init_shift_var(m):
+    m_hat = m // 2
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/keras/quantizer/gumbel_rounding/base_gumbel_rounding.py:38]: Comments and type hints.

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

**Review by reuvenperetz — APPROVED** [review]
